packages/core/contracts/oracle/implementation/OptimisticOracle.sol

```diff
@@ -0,0 +1,368 @@
+        int256 proposedPrice
+    );
+    event DisputePrice(
+        address indexed requester,
+        address indexed proposer,
+        address indexed disputer,
+        bytes32 identifier,
+        uint256 timestamp
+    );
+    event Settle(
+        address indexed requester,
+        address indexed proposer,
+        address indexed disputer,
+        bytes32 identifier,
+        uint256 timestamp,
+        int256 price
+    );
+
+    enum State {
+        Invalid, // Never requested.
+        Requested, // Requested, no other actions taken.
+        Proposed, // Proposed, but not expired or disputed yet.
+        Expired, // Proposed, not disputed, past liveness.
+        Disputed, // Disputed, but no DVM price returned yet.
+        Resolved, // Disputed and DVM price is available.
+        Settled // Final price has been set in the contract (can get here from Expired or Resolved).
+    }
+
+    struct Request {
+        address proposer;
+        address disputer;
+        IERC20 currency;
+        bool settled;
+        bool refundOnDispute;
+        int256 proposedPrice;
+        int256 resolvedPrice;
+        uint256 expirationTime;
+        uint256 reward;
+        uint256 finalFee;
+        uint256 bond;
+        uint256 customLiveness;
+    }
+
+    mapping(bytes32 => Request) public requests;
+
+    FinderInterface public finder;
+    uint256 public defaultLiveness;
+
+    constructor(
+        uint256 _liveness,
+        address _finderAddress,
+        address _timerAddress
+    ) public Testable(_timerAddress) {
+        finder = FinderInterface(_finderAddress);
+        _validateLiveness(_liveness);
+        defaultLiveness = _liveness;
+    }
+
+    function requestPrice(
+        bytes32 identifier,
+        uint256 timestamp,
+        IERC20 currency,
+        uint256 reward
+    ) external nonReentrant() returns (uint256 totalBond) {
+        require(getState(msg.sender, identifier, timestamp) == State.Invalid, "requestPrice: Invalid");
+        require(_getIdentifierWhitelist().isIdentifierSupported(identifier), "Unsupported identifier");
+        require(_getCollateralWhitelist().isOnWhitelist(address(currency)), "Unsupported currency");
+        uint256 finalFee = _getStore().computeFinalFee(address(currency)).rawValue;
+        requests[_getId(msg.sender, identifier, timestamp)] = Request({
+            proposer: address(0),
+            disputer: address(0),
+            currency: currency,
+            settled: false,
+            refundOnDispute: false,
+            proposedPrice: 0,
+            resolvedPrice: 0,
+            expirationTime: 0,
+            reward: reward,
+            finalFee: finalFee,
+            bond: finalFee,
+            customLiveness: 0
+        });
+
+        if (reward > 0) {
+            currency.safeTransferFrom(msg.sender, address(this), reward);
+        }
+
+        emit RequestPrice(msg.sender, identifier, timestamp);
+
+        return finalFee.mul(2);
```

Comment:
I don't see this return val being used in the FundingRateApplier

Reply:
Yeah, the funding rate applier changes the bond in a subsequent call, and the return value from that call is stored (since it would override this value anyway). However, I think returning the bond is generally useful for callers. In fact, I think I'm going to add it to propose and dispute too for good measure.

Thoughts?

Reply:
Yeah that's reasonable